decision-records/2025-03-20--node-exporter-otelcol-adr.md

```diff
@@ -0,0 +1,396 @@
+
+```mermaid
+flowchart LR
+subgraph host
+    subgraph principal-charm1
+        principal-charm-app1
+    end
+    subgraph principal-charm2
+        principal-charm-app2
+    end
+    subgraph cos-collector-charm
+    direction BT
+        direction BT
+        port9100["port: 9100"]
+        node-exporter1["node-exporter"]
+        node-exporter1 --> port9100
+
+        port4317["port: 4317"]
+        port4318["port: 4318"]
+        otelcol1["otelcol"]
+        otelcol1 --> port4317
+        otelcol1 --> port4318
+    end
+    cos-collector-charm --->|"subordinate"| principal-charm1
+    cos-collector-charm --->|"subordinate"| principal-charm2
+end
+```
+
+If the idea is so simple, why wouldn't we implement it? Well, the answer is that from a charm perspective is not that simple. In fact in `grafana-agent` charm it is not implemented, and because of this we have [some weird situations.](https://discourse.charmhub.io/t/one-grafana-agent-charm-to-rule-them-all/16014).
+
+### Some considerations to be taken into account when implementing this solution
+
+
+Everytime `cos-collector` subordinate charm is related to a principal charm, `cos-collector` charm must:
+
+* Verify whether `otelcol` and `node-exporter` snaps are already installed or not to avoid trying to install them again.
+* Merge the `otelcol` configuration resulting from the established relationship with any previously existing configuration.
+
+
+When a relation between `cos-collector` and a principal charm is removed, the `cos-collector` charm must:
+
+* Remove from the `otelcol` config file the configuration resulting from the departing relation.
+* Uninstall `otelcol` and `node-exporter` snaps only if there is no other `cos-collector` charm deployed in the same host.
+
+
+#### Questions and doubts about this approach
+
+
+* When we think of a charm, we expect it to manage the lifecycle of an application, from installation, configuration, scaling to relationships. However, with this approach, multiple instances of `cos-collector` charm will handle a single binary and a single configuration file.
+
+```mermaid
+flowchart LR
+subgraph host
+    direction TB
+
+    subgraph cos-collector-charm
+        direction TB
+        charm-code
+        subgraph "Binaries and config"
+            node-exporter1["node-exporter binary"]
+            node-exporter-config["node-exporter configuration"]
+
+            otelcol["otelcol binary"]
+            otelcol-config["otelcol configuration"]
+        end
+        charm-code --> otelcol-config
+        charm-code --> node-exporter-config
+    end
+
+    cos-collector-charm --->|"subordinate"| principal-charm1
+    cos-collector-charm --->|"subordinate"| principal-charm2
+end
+```
+
+## Decision
+
+Based on the previous analysis, the decision is to follow the second approach: *"Only one `otelcol` + `node-exporter` binaries per `cos-collector` charm (and per principal charm and host)"*.
```

Comment:
Not sure what you mean here. Personally I am leaning towards:
- node-exporter and otelcol are separate snaps;
- the charm can still be called opentelemetry-collector.

Reply:
We normally name the charms with the name of the workload they manage.
Since this charm will manage both `otelcol` and `node-exporter` that why I've suggested a new name... Anyway, we can take the decision to name it just `opentelemetry-collector` because is the "main" snap.